src/hotspot/share/opto/graphInvariants.cpp

```diff
@@ -0,0 +1,617 @@
+
+struct NodeClass : Pattern {
+  explicit NodeClass(bool (Node::*type_check)() const) : _type_check(type_check) {}
+  bool check(const Node* center, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    if (!(center->*_type_check)()) {
+      ss.print_cr("Unexpected type: %s.", center->Name());
+      return false;
+    }
+    return true;
+  }
+  bool (Node::*_type_check)() const;
+};
+
+struct HasNOutputs : Pattern {
+  explicit HasNOutputs(uint expect_outcnt) : _expect_outcnt(expect_outcnt) {}
+  bool check(const Node* center, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    if (center->outcnt() != _expect_outcnt) {
+      ss.print_cr("Unexpected number of outputs. Expected: %d, found: %d.", _expect_outcnt, center->outcnt());
+      for (DUIterator_Fast imax, i = center->fast_outs(imax); i < imax; i++) {
+        Node* out = center->fast_out(i);
+        ss.print("  ");
+        out->dump("\n", false, &ss);
+      }
+      return false;
+    }
+    return true;
+  }
+  const uint _expect_outcnt;
+};
+
+struct AtSingleOutputOfType : Pattern {
+  AtSingleOutputOfType(bool (Node::*type_check)() const, const Pattern* pattern) : _type_check(type_check), _pattern(pattern) {
+  }
+
+  bool check(const Node* center, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    Node_List outputs_of_correct_type;
+    for (DUIterator_Fast imax, i = center->fast_outs(imax); i < imax; i++) {
+      Node* out = center->fast_out(i);
+      if ((out->*_type_check)()) {
+        outputs_of_correct_type.push(out);
+      }
+    }
+    if (outputs_of_correct_type.size() != 1) {
+      ss.print_cr("Non-unique output of expected type. Found: %d.", outputs_of_correct_type.size());
+      for (uint i = 0; i < outputs_of_correct_type.size(); ++i) {
+        outputs_of_correct_type.at(i)->dump("\n", false, &ss);
+      }
+      return false;
+    }
+    bool result = _pattern->check(outputs_of_correct_type.at(0), steps, path, ss);
+    if (!result) {
+      steps.push(outputs_of_correct_type.at(0));
+      path.push(LocalGraphInvariant::OutputStep);
+    }
+    return result;
+  }
+  bool (Node::*_type_check)() const;
+  const Pattern* const _pattern;
+};
+
+struct PatternBasedCheck : LocalGraphInvariant {
+  const Pattern* const _pattern;
+  explicit PatternBasedCheck(const Pattern* pattern) : _pattern(pattern) {}
+  CheckResult check(const Node* center, LazyReachableCFGNodes& reachable_cfg_nodes, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    return _pattern->check(center, steps, path, ss) ? CheckResult::VALID : CheckResult::FAILED;
+  }
+};
+
+struct IfProjections : PatternBasedCheck {
+  IfProjections()
+      : PatternBasedCheck(
+            And::make(
+                new HasNOutputs(2),
+                new AtSingleOutputOfType(&Node::is_IfTrue, new True()),
+                new AtSingleOutputOfType(&Node::is_IfFalse, new True()))) {
```

Comment:
I would suggest that you append the word `Pattern` to all `Patterns` - at least in most cases this will make it a bit easier to see what you have at the use-site. I'm looking at `new True()` and wonder what might be passed here... if it was called `TruePattern`, it would be immediately clear.

Reply:
> I would suggest that you append the word Pattern to all Patterns - at least in most cases this will make it a bit easier to see what you have at the use-site. I'm looking at new True() and wonder what might be passed here... if it was called TruePattern, it would be immediately clear.

For `True` alright, no strong opinion. Could also be `TrivialPattern` or so. For everything else, that looks very verbose and hurts readability a lot, and I think readability is very important. I think the other patterns are pretty understandable: for instance, I don't see how `HasAtLeastNInputsPattern` would really help compared to `HasAtLeastNInputs`, it seems just like bloat my brain will have to strip.